api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,629 @@
+		hasRoom := false
+
+		// first, sort by memory so we pack tasks by memory as tightly as possible
+		sortProvidersByMemory(providers)
+
+		// next, place any unused providers in the back of the list
+		// that way, we can can delete them if we avoid placing any tasks in them
+		sortProvidersByUsage(providers)
+
+		for _, provider := range providers {
+
+			if hasResourcesFor(consumer, provider) {
+				hasRoom = true
+				subtractResourcesFor(consumer, provider)
+				break
+			}
+		}
+
+		if !hasRoom {
+			newProvider, err := e.calculateNewProvider(clusterName)
+			if err != nil {
+				return nil, err
+			}
+
+			if !hasResourcesFor(consumer, newProvider) {
+				text := fmt.Sprintf("Resource '%s' cannot fit into an empty provider!", consumer.ID)
+				text += "\nThe instance size in your environment is too small to run this resource."
+				text += "\nPlease increase the instance size for your environment"
+				err := fmt.Errorf(text)
+				errs = append(errs, err)
+				continue
+			}
+
+			subtractResourcesFor(consumer, newProvider)
+			providers = append(providers, newProvider)
+		}
+	}
+
+	// check if we need to scale down
+	unusedProviders := []*models.ResourceProvider{}
+	for _, provider := range providers {
+		if !provider.InUse {
+			unusedProviders = append(unusedProviders, provider)
+		}
+	}
+
+	desiredScale := len(providers) - len(unusedProviders)
+
+	actualScale, err := e.scaleTo(clusterName, desiredScale, unusedProviders)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	info := &models.ScalerRunInfo{
+		EnvironmentID:           clusterName,
+		PendingResources:        resourceConsumerModels(consumers),
+		ResourceProviders:       resourceProviderModels(providers),
+		ScaleBeforeRun:          scaleBeforeRun,
+		DesiredScaleAfterRun:    desiredScale,
+		ActualScaleAfterRun:     actualScale,
+		UnusedResourceProviders: len(unusedProviders),
+	}
+
+	return info, errors.MultiError(errs)
+}
+
+func resourceConsumerModels(consumers []models.ResourceConsumer) []models.ResourceConsumer {
+	consumerModels := make([]models.ResourceConsumer, len(consumers))
+	for i, consumer := range consumers {
+		newConsumer := models.ResourceConsumer{
+			ID:     consumer.ID,
+			Memory: consumer.Memory,
+			Ports:  consumer.Ports,
+		}
+		consumerModels[i] = newConsumer
+	}
+
+	return consumerModels
+}
+
+func resourceProviderModels(providers []*models.ResourceProvider) []models.ResourceProvider {
+	providerModels := make([]models.ResourceProvider, len(providers))
+	for i, provider := range providers {
+		newProvider := models.ResourceProvider{
+			ID:              provider.ID,
+			InUse:           provider.InUse,
+			UsedPorts:       provider.UsedPorts,
+			AvailableMemory: provider.AvailableMemory,
+		}
+		providerModels[i] = newProvider
+	}
+
+	return providerModels
+}
+
+func (e *EnvironmentScaler) getResourceProviders(clusterName string) ([]*models.ResourceProvider, error) {
+	listContainerInstancesInput := &ecs.ListContainerInstancesInput{}
```

Comment:
[nit] I know we have a guideline of not calling helper methods from other helper methods but it sure would be cleaner in my opinion to wrap the aws calls in their own method. So getResourceProviders would look something like

```go
func (e *EnvironmentScaler) getResourceProviders(clusterName string) ([]*models.ResourceProvider, error) {
   clusterInstances, err := getContainerInstances(clusterName)
   if err != nil ...

```